### PR TITLE
[terraformer-v2] Implement terraformer v2 golang wrapper

### PR DESCRIPTION
**Pull request by timebertt**
**What this PR does / why we need it**:
This PR rewrites terraformer in go.

Now terraformer is basically a go program wrapping terraform, which delegates commands and signals to terraform itself and watches the state file to continuously safe the state as soon as it gets written to disk.

⚠️ This is not backwards-compatible as the docker image command changes and the parameters are now passed in via flags instead of env vars (see examples for details).

/area quality
/area robustness

**Which issue(s) this PR fixes**:
Part of #40 

**Special notes for your reviewer**:
This PR is split up into multiple commits to make reviews at least a bit easier. Though, commits are not completely correct independently and might be a bit out of order (we can squash them later on before merging).

The base branch **is not master**, but rather a feature branch. As this PR brings a lot of changes in different repos, we might have to do some smaller adaptations later on, when trying to use it in all the extensions. So let's get this in as a first step, then try to adapt everything else and only merge it on master, when we are sure, that we can release everything together.

TODOs:
- ✅ add unit tests for a few remaining cases
- ✅ complete binary e2e tests
- ✅ clarify all `// TODO`s in code
- ✅ Figure out, how to adapt CI config to the changes

Follow-up PRs:
- 👀 Add pod e2e tests (test-machinery style integration test) with real terraform IaaS calls (#51)
- ✅ [g/g/ext] Adapt terraformer library to breaking changes: https://github.com/gardener/gardener/pull/3034 (released with [g/g@v1.12.0](https://github.com/gardener/gardener/releases/tag/v1.12.0))
- 🚧 [g/providers] Adapt provider extensions to breaking changes (doesn't require many changes, see gcp for a reference implementation: https://github.com/gardener/gardener-extension-provider-gcp/pull/190)

**Release note**:

```action developer
Terraformer version `v2` now requires to pass in ConfigMap and Secret names via command line flags instead of environment variables and the base command of the docker image has changed to `/terraformer`. Please see the [examples](https://github.com/gardener/terraformer/tree/master/example) for more details.
```
```action developer
If your provider extension is deploying terraformer Pods via gardener's [terraformer library](https://github.com/gardener/gardener/tree/master/extensions/pkg/terraformer), please make sure, that you use at least `gardener/gardener@v1.12.0` and set `terraformer.UseV2(true)` in order to deploy a PodSpec, that is compatible with terraformer `v2`.
```
```noteworthy operator
Terraformer was rewritten in go and now watches the terraform state file in order to continuously update the state ConfigMap to not lose any relevant infrastructure state.
```
```improvement developer
You can use `make start-dev-container` to start a docker container which can run terraformer and tests in an isolated environment for development and testing. Also `make start` will run terraformer commands in such a development container.
```

### CODE REVIEW

File: go.mod
Comment (rfranzke) [inline go.mod:3]:
```suggestion
go 1.15
```
?

File: pkg/terraformer/paths.go
Comment (rfranzke) [inline pkg/terraformer/paths.go]: As tested, I guess it's not needed anymore

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:30]: doc string missing

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:52]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:56]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:65]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:80]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:87]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:91]: doc string

File: pkg/terraformer/store.go
Comment (rfranzke) [inline pkg/terraformer/store.go:100]: doc string

File: pkg/terraformer/terraformer.go
Comment (rfranzke) [inline pkg/terraformer/terraformer.go:66]:
```suggestion
	}
  t.client = c
```

File: pkg/terraformer/terraformer.go
Comment (rfranzke) [inline pkg/terraformer/terraformer.go:82]:
```suggestion
	sigCh := make(chan os.Signal, 1)
```

File: pkg/terraformer/terraformer.go
Comment (rfranzke) [inline pkg/terraformer/terraformer.go]: see above, can be removed

File: pkg/terraformer/types.go
Comment (rfranzke) [inline pkg/terraformer/types.go:34]: doc strings

File: pkg/terraformer/types.go
Comment (rfranzke) [inline pkg/terraformer/types.go:53]: doc string

File: pkg/utils/exit_code.go
Comment (rfranzke) [inline pkg/utils/exit_code.go:34]: doc string

File: test/utils/constants.go
Comment (rfranzke) [inline test/utils/constants.go:22]: doc strings

File: pkg/terraformer/config.go
Comment (rfranzke) [inline pkg/terraformer/config.go:270]:
```suggestion
  return watcher.Add(t.paths.StatePath)
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Very nice PR, really appreciate it. I don't have much valuable comments to the code itself, I will need some time to play with it and try it out 👀
I've just left a few minor comments for now.

**Comment by rfranzke** [comment]
Ah, and can we replace the license headers with the new REUSE, please?

File: go.mod
Comment (timebertt) [reply go.mod]: Done

**Comment by timebertt** [comment]
I've now configured the CI properly for this branch and added the `verify` step.
Also applied @rfranzke's suggestions.

/ready

**Comment by timebertt** [comment]
> Ah, and can we replace the license headers with the new REUSE, please?

Done.

Also copied `g/g/test/framework/cleanup.go` as it's not isolated from other dependencies and therefore brought in a lot of unneeded vendored stuff.

**Review by ialidzhikov — COMMENTED** [review]
Very nice, I like it! I also don't have much valuable comments from my initial "go through".
I tested terraformer@v2 with provider-gcp and provider-aws (also with infra with 3 zones -> which results in 58 terraform resources in total) and it was behaving as expected.

File: .golangci.yaml
Comment (timebertt) [reply .golangci.yaml:13]: Good catch. No, we don't use any deprecated variables/funcs here.
Removed it.

File: pkg/terraformer/terraformer.go
Comment (timebertt) [reply pkg/terraformer/terraformer.go:100]: No, currently, there is no backoff for any update, as I wasn't sure, how to properly do the retries. I still have to add it.

As discussed, I will probably add the retry only for the final state update and don't retry the other ones.
After a quick chat with @rfranzke, I think a retry strategy that makes sense would be:
- retry on any error with an exponential backoff
- after that, keep retrying with a fixed interval (last duration of backoff)
- timeout after ~1h or so.

WDYT?

Comment (ialidzhikov) [reply pkg/terraformer/terraformer.go:100]: Makes sense. Thank you!

**Comment by timebertt** [comment]
I've pushed a few new commits with the following changes:
- Implement retries using queue and worker approach (as discussed) 684220c
  - also cleans up large defer funcs in `execute` to make it more readable
- Vendor g/g@v1.12.4 788d698
  - to get the newest improvements in the hack scripts
- Split up config.go, add more unit tests for state handling 852f85b
- Add unit tests for exit code and signal handling 653b6fe

Thus:

> add unit tests for a few remaining cases

is finished. (test cov is at over 90%)
Also,

>  clarify all `// TODOs` in code

is done. Retries are now properly implemented, while making sure, that conflicting state changes don't overwrite each other (only one worker reads from the state update queue).

/remove status/author-action needs/changes
/unassign

**Comment by timebertt** [comment]
/invite @rfranzke @ialidzhikov @timuthy 
This is now ready for another round of review 😄 

I will check the CI failure tomorrow.

If you agree, we can also postpone the last remaining task (to finish the binary e2e tests) to another PR on this feature branch. WDYT?

**Comment by timebertt** [comment]
Fixed the build by removing the test cases, that were provoking errors in `os.Mkdir` or `os.Open` by changing some permissions on the fs.

This approach doesn't work in the dev-container and in the pipeline, as `go test` is executed as `root` there. Anyways, the error handling is trivial in that cases, so the tests were somewhat pointless.

**Comment by timebertt** [comment]
I continued working on the binary e2e tests. Now only a few test cases are missing :)
Also, I optimized the runtime of the existing unit tests.

File: cmd/terraformer/app/terraformer.go
Comment (timuthy) [inline cmd/terraformer/app/terraformer.go:115]: Can you explain when and how this is used?

**Comment by timebertt** [comment]
> complete binary e2e tests

Is ✅ now, I added the few missing test cases for the file watcher, signal handling and exit code handling.
Also, terraform output is now completely redirected to `stderr` to keep it consistent with all other logs.

With this, no more TODOs for this PR, it is completely ready now.
Please go ahead with the reviews 😉

**Review by timuthy — COMMENTED** [review]
Very nice!

I left one question and have the following suggestions:
- Can you please adapt `README.md` wrt your changes?
- Should we move `scripts/fetch-providers` -> `hack`?

File: cmd/terraformer/app/terraformer.go
Comment (timebertt) [reply cmd/terraformer/app/terraformer.go:115]: It is used instead of the default cobra usage template, i.e. if `terraformer help` or `terraformer [command] -h` is executed.

It differs only slightly from the default template. E.g. for `terraformer help` it puts out 
```
...
Usage:
  terraformer [command] [flags]
...
```
instead of 
```
...
Usage:
  terraformer [flags]
  terraformer [command]
...
```

So as the name suggests, this is only a minor cosmetic enhancement 😄

Comment (timuthy) [reply cmd/terraformer/app/terraformer.go:115]: Okay, thanks for explaining. I especially missed the `Examples:` section being printed in the help but that's probably because you don't have any examples configured. Never mind, I mainly asked out of curiosity reasons :)

Comment (timebertt) [reply cmd/terraformer/app/terraformer.go:117]: Specifying an example is probably a good idea anyways, so I added one for each command :)

**Comment by timebertt** [comment]
> * Can you please adapt `README.md` wrt your changes?

Oh yes, totally missed that. Done now :)

> * Should we move `scripts/fetch-providers` -> `hack`?

Yep. Also done.
PTAL :)

**Comment by timebertt** [comment]
One question came to my mind:
What should the exit code of terraformer be, if it was interrupted?
Currently, if it is interrupted during terraform execution, it will exit with the same error code as terraform (except when terraform exits successfully, but terraformer can't store the state).

**Review by rfranzke — APPROVED** [review]
/lgtm

No further comments as of now. I suggest to proceed and merge this PR maybe end of this week if nobody objects, and then try it out with the GCP extension as discussed.

**Comment by rfranzke** [comment]
> What should the exit code of terraformer be, if it was interrupted?
Currently, if it is interrupted during terraform execution, it will exit with the same error code as terraform (except when terraform exits successfully, but terraformer can't store the state).

Sounds good to me at the moment. Do you have concerns?

**Comment by timebertt** [comment]
> > What should the exit code of terraformer be, if it was interrupted?
> > Currently, if it is interrupted during terraform execution, it will exit with the same error code as terraform (except when terraform exits successfully, but terraformer can't store the state).
> 
> Sounds good to me at the moment. Do you have concerns?

No, I don't. Just wanted to raise the question and see if someone has concerns...

**Review by danielfoehrKn — COMMENTED** [review]
This review is late. My bad, I should have marked this PR in review by me.

In any case, I only have one question, the rest looks good to me.

File: pkg/terraformer/terraformer.go
Comment (timebertt) [reply pkg/terraformer/terraformer.go:172]: Yeah, I was also thinking about redirecting the terraformer output to the log or prefixing it somehow.
I did it this way, because we also had it like this before. But we can definitely change this later on, if we think it makes sense.
Should we keep it like this for the time being.

Comment (timebertt) [reply pkg/terraformer/terraformer.go:172]: Now I also remember the full reason for not doing it like that:
The terraformer pod logs are fetched by the provider extensions and they put parts of the output into the `lastError` fields (which in turn is used to calculate error codes). 
So I wanted to be a bit more cautious here, as this might 1) look really bad in the dashboard (if the logs JSON fields are not stripped) and 2) might even break our error code detection in an unforeseeable way.
So, we can definitely revise that idea, but I guess we will also need some changes in `g/g/ext/terraformer` with it, too.

**Pull request merged**
